### PR TITLE
Extension storage sync and updated query params and files

### LINKED ISSUE #121: FeatureRequest: Additional gist integration parameters
At the moment its possible to do something like https://chriseth.github.io/browser-solidity/?gist=f9aa1278818ef2f022fd , which is extremely helpful. But what would even be better is the ability to also set the "optimizer" and "compiler versions" via querystring. So it should end up something like were we can past both the compiler version and optimization settings

https://chriseth.github.io/browser-solidity/?gist=f9aa1278818ef2f022fd&compiler=v.x.x.x.x&optimizer=1

### PR DESCRIPTION
**Pull request by d11e9**
Changes visible/testable at: http://d11e9.github.io/browser-solidity/
- [x] Use chrome.storage.sync to sync files between browsers (extension only)
- [x] Update file tab UI to allow many files and scrollable
- [x] Add version and optimize to URL to support reproduceability fixes: #121 

**TODO**:
- [x] Include version and optimize flag in gist output on publish
- [ ] Make undo stack usable and maintain between file tabs
- [ ] Better UI for cloud/local overwritting

### CODE REVIEW

**Comment by d11e9** [comment]
Worth noting that this changes usage of url params from `?param=value&x=y` to `#param=value&x=y` but does handle legacy urls by converting to `#` params on load

**Comment by d11e9** [comment]
Rebased against upstream

**Comment by chriseth** [comment]
Sorry,created a conflict. Is it possible to do the two todo items that are left as a separate PR?

**Comment by d11e9** [comment]
@chriseth if you're happy merging whats done so far then sure, I can continue in another PR. I don't believe there are regressions in its current state. Only room for improvement.

**Comment by d11e9** [comment]
Will rebase asap.

**Comment by chriseth** [comment]
@d11e9 did you use "rebase" as "synonym" for "merge"? Please do an actual rebase as merge conflict resoultion cannot be reviewed in github.

**Comment by d11e9** [comment]
Had not meant it as a synonym but ended up doing a merge when I got a chance, will rebase properly ☺️

**Comment by d11e9** [comment]
@chriseth rebased ;)

File: assets/js/app.js [hunk 1]
```diff
@@ -43,23 +87,72 @@
...
 					}
 				});
 			}
 
 
+			// ----------------- storage --------------------
+
+			function syncStorage() {
+
+				if (!chrome || !chrome.storage || !chrome.storage.sync) return;
```
Comment (chriseth) [inline assets/js/app.js]: This still breaks on firefox ("reference error"). I think a fix would be
`if (typeof chrome === 'undefined' || !chrome || ...`

**Comment by axic** [comment]
@chriseth @d11e9 I have a strange issue since merging this. When there are multiple files (tabs) open, it won't compile (or refresh the right side?) when switching between tabs. Only the first tab is compiled.

**Comment by chriseth** [comment]
Should be fixed now by https://github.com/chriseth/browser-solidity/pull/148

**Comment by axic** [comment]
Works here!

**Pull request merged**